### ISSUE TITLE
How createTexture can change 1D 256x1 rgba data to 2D 16 x 16 rgba data? Non resizing question

### ISSUE THREAD

**Issue by wood1986** [greggman/twgl.js#166]
I have a gradient png whose width and height are 256 and 1 respectively. When I use [`createTexture`](https://twgljs.org/docs/module-twgl.html#.createTexture) to create a texture, [`TextureOptions`](https://twgljs.org/docs/module-twgl.html#.TextureOptions) will ignore width and height.

How can I move the pixel data in a png from 256x1 to 16x16? I am not talking about resizing. I want to do something that I pass an UInt8Array as the `src` in [`TextureOptions`](https://twgljs.org/docs/module-twgl.html#.TextureOptions) and then specify the width and height, the texture change

from 1D
```
RGBA(0,0), RGBA(1,0), ... RGBA(255,0) // which is the 256 x 1 texture
```
to 2D
```
RGBA(0,0), RGBA(1,0), ... RGBA(15,0) //  
RGBA(0,1), RGBA(1,1), ... RGBA(15,0) // This will be equivalent to RGBA(16,0), RGBA(31,0) in 256 x 1 texture.
.
.
.
RGBA(0,15), RGBA(1,15), ... RGBA(15,15) // This will be equivalent to RGBA(239,0), RGBA(255,0) in 256 x 1 texture.
```

In short, how to get the [`createTexture`](https://twgljs.org/docs/module-twgl.html#.createTexture) to respect `width` and `height` when `src` is non-Array

**Comment by greggman** [comment]
It should be enough to pass in width or height or both

```
const tex = twgl.createTexture(gl, {
  width: 256,
  src: someUint8ArrayWith256x4Values,
});
```

This should also work

```
const tex = twgl.createTexture(gl, {
  height: 1,
  src: someUint8ArrayWith256x4Values,
});
```

Basically it looks at the format/type to figure out the number of bytes per pixel and then divides the number of pixels by width or height, which ever is specified or it does `Math.sqrt` if neither are specified or it does nothing of both are specified. 

format defaults to `RGBA`, type defaults to `UNSIGNED_BYTE` but is derived from `src` if it's a typedarray

**Comment by greggman** [comment]
If source is not an array, like if it's an image, then you can't do it directly via twgl. The functions that take an non-array in WebGL don't provide those options and neither does twgl.

If you don't care about alpha you load the image yourself, copy it to a png like this

```
function loadImage(url) {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.onload = () => resolve(img);
    img.onerror = reject;
    img.src = url;
  });
}

async load() {
  const img = loadImage('url/for/image');
  const canvas = document.createElement('canvas');
  canvas.width = img.width;
  canvas.height = img.height;
  const ctx = canvas.getContext('2d');
  ctx.drawImage(img, 0, 0);
  const imageData = ctx.getImageData(0, 0, img.width, img.height);
  const tex = twgl.createTexture(gl, { width: 16, src: imageData.data, });
}
```

**Comment by wood1986** [comment]
What if I cannot use canvas?

**Comment by greggman** [comment]
3 ideas?

(1) use a js png decoder like https://github.com/photopea/UPNG.js/

(2) don't use a png. Just a raw binary file

(3) In WebGL2 (not WebGL1) there are a bunch of [`pixelStorei` settings](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/pixelStorei) you can set so you can call `gl.texImage2D(target, level = 0, gl.RGBA8, width = 16, height = 16, 0, gl.RGBA, gl.UNSIGNED_BYTE, some256x1Image)` . In fact it might just work as is with no settings given 16x16 = 256x1

Maybe I can change twgl so if it's WebGL2 and you pass in an image and you pass a width or height that it tries to use those features if not using them directly it calls the version of texImage2D that takes an image and a width and height (WebGL1 has no such entry point)

**Comment by wood1986** [comment]
@greggman You read my mind because I plan to give UPNG a trial after researching lots PNG decoder.

I will try pixelStorei as well.


> Maybe I can change twgl so if it's WebGL2 and you pass in an image and you pass a width or height that it tries to use those features if not using them directly it calls the version of texImage2D that takes an image and a width and height (WebGL1 has no such entry point)

+1 to that

**Comment by greggman** [comment]
Hmm, apparently pixelStorei etc won't work for this use case. Or else I'm mis-understanding something. So yea, you'll need to unpack via a png decoder.
